### ISSUE TITLE
@CartesianCsvSource for @CartesianProductTest

### ISSUE THREAD

**Issue by scordio** [junit-pioneer/junit-pioneer#416]
Following #409, it would be nice to have `@CartesianCsvSource` which mimics `@CsvSource`.

Example:

```java
@CartesianProductTest
@CartesianValueSource(ints = { 1, 2 })
@CartesianCsvSource({
  "A, B",
  "C, D",
  "E, F",
})
void moreParams(int i, String s1, String s2) {
  // some test code
}
```
should generate 6 tests:

| Test # | i | s1 | s2 |
|--------|---|----|----|
| 1      | 1 | A  | B  |
| 2      | 2 | A  | B  |
| 3      | 1 | C  | D  |
| 4      | 2 | C  | D  |
| 5      | 1 | E  | F  |
| 6      | 2 | E  | F  |

**Comment by mizosoft** [comment]
That would be really nice. I was about to open a similar issue as I've just realized `@CartesianProductTest` doesn't work with JUnit's param annotations when I tried it with `@CsvFileSource`. I'm wondering though, why there exists such a limitation? I think it would be nicer if one could use JUnit's standard param annotations instead of the the practically similar `@CartesianX` ones.

**Comment by Michael1993** [comment]
> I'm wondering though, why there exists such a limitation? I think it would be nicer if one could use JUnit's standard param annotations instead of the the practically similar @CartesianX ones.

@mizosoft There is a bit of a clash.

When you create a `@ParameterizedTest`, you supply a single (!) `@ArgumentsSource` annotation where you define every test case. This is also the reason why annotations annotated with `@ArgumentsSource` (e.g.: `@CsvSource` or `@ValueSource`) are not `@Repeatable`.

This is different from `@CartesianProductTest`, where you want to define argument sets for each parameter individually and let the extension define the test cases. This is an entirely different approach from `@ParameterizedTest`.

We cannot support the annotations used by `@ParameterizedTest` because we would need them to be `@Repeatable`.
They (*probably*) never will be repeatable because their design philosophy is different from what we need.

**Comment by Michael1993** [comment]
My above comment is also what I consider a major challenge when trying to implement `@CartesianCsvSource` - determining which `ArgumentsProvider` should be invoked for the different arguments.

**Comment by scordio** [comment]
@Michael1993 I'm happy to take this up and see what we can get

**Comment by mizosoft** [comment]
@Michael1993 I see your point. Thanks for clarifying! I'm still wondering however if these annotations can still be honored by `@CartesianProductTest` despite them being unrepeatable. I don't see why something like this won't work:

```java
  @CartesianProductTest
  @EnumSource(MyEnum.class)
  @ValueSource(ints = {1, 2, 3})
  void bob(MyEnum myEnum, int myInt) {
    // ...
  }
```

I think that you'd normally have one `XSource` for each `X`, so you don't need them to be repeatable. Also, I've noticed that JUnit actually allows multiple sources on the same parameterized method (not the same source though). It's just that it interprets them differently by forming one linear set from of all the sources instead of their product (so for example you can have `@NullSource` and `@EmptySource` and `@ValueSource` in the same method).

**Comment by scordio** [comment]
I have preliminary feedback about this topic. The simple copy-paste from JUnit does not work out of the box because the `CartesianProductTestExtension` flattens the parameters returned by the provider. See:

https://github.com/junit-pioneer/junit-pioneer/blob/master/src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java#L137

In this case, the provider is returning an array of values per execution, and all together should be treated as a single element in the cartesian product but since the array is flattened all the values are used independently.

I'll continue looking into it and propose a solution.

**Comment by scordio** [comment]
@Michael1993 I plan to work on this Soon™ 🙂 is there any refactoring I should wait for (e.g., https://github.com/junit-pioneer/junit-pioneer/issues/447#issuecomment-814036571) to minimize conflicts?

**Comment by Michael1993** [comment]
After #487, we need to think of a new approach to this extension. Perhaps if `@CartesianCsvSource` provided values to all parameters?

Personally, I think instead of this:
```java
@CartesianProductTest
@CartesianValueSource(ints = { 1, 2 })
@CartesianCsvSource({
  "A, B",
  "C, D",
  "E, F",
})
void moreParams(int i, String s1, String s2) {
  // some test code
}
```

Now we can have this:
```java
@CartesianProductTest
@CartesianCsvSource({
  "1, 2",
  "A, C, E",
  "B, D, F"
})
void moreParams(int i, String s1, String s2) {
  // some test code
}
```

What's your thoughts @scordio?

**Comment by scordio** [comment]
If I understand your proposal correctly, each line would provide all the possible values for each parameter. So, with:

> ```java
> @CartesianProductTest
> @CartesianCsvSource({
>   "1, 2",
>   "A, C, E",
>   "B, D, F"
> })
> void moreParams(int i, String s1, String s2) {
>   // some test code
> }
> ```

we would have 2 x 3 x 3 = 18 tests, while my original proposal was targeting only 2 x 3 = 6 tests, where the "2" comes from values { 1, 2 } for the `i` parameter, and the "3" comes from the three pairs { (A, B), (C, D), (E, F) } for parameters `s1` and `s2`.

I think the new `@CartesianValueSource` on parameter level would be more understandable to target your example and also it might be easier for the users not to mix up the behavior of `@CartesianCsvSource` with `@CsvSource` of JUnit, where each CSV string provides values for all the parameters (sort of "horizontal" vs. "vertical" expansion).

Sticking to the original behavior I proposed, my feeling is that `@CartesianCsvSource` has to stay on a method level because it would cover more than one parameter. The challenge is how to make it work nicely with the other sources which will be on parameter level after #487.

Let's say we have:

> ```java
> @CartesianProductTest
> @CartesianCsvSource({
>   "A, B",
>   "C, D",
>   "E, F",
> })
> void moreParams(@CartesianValueSource(ints = { 1, 2 }) int i, String s1, String s2) {
>   // some test code
> }
> ```

Can we have a very precise mapping to the target parameters for `@CartesianCsvSource`? The example above might be easy, we could simply exclude all the parameters having a source annotation, but I'm not sure if there can be corner cases that cannot be handled.

One more example:

> ```java
> @CartesianProductTest
> @CartesianCsvSource({
>   "A, B",
>   "C, D",
>   "E, F",
> })
> void moreParams(String s1, @CartesianValueSource(ints = { 1, 2 }) int i, String s2) {
>   // some test code
> }
> ```

Would / should this be possible? (Expectation: 3 x 2 = 6 tests)

Also, should `@CartesianCsvSource` be repeatable? (which might cause #447 again if there is more than one repeatable annotation type relevant for @CartesianProductTest)

For example:

> ```java
> @CartesianProductTest
> @CartesianCsvSource({
>   "A, B",
>   "C, D",
>   "E, F",
> })
> @CartesianCsvSource({
>   "G, H",
>   "I, J",
> })
> void moreParams(String s1, String s2, String s3, String s4) {
>   // some test code
> }
> ```

Is it supposed to work? (Expectation: 3 x 2 = 6 tests)

**Comment by nipafx** [comment]
Can you two take another look at this and come to a decision whether something could be done here given the new approach to Cartesian tests and what that "something" could look like?
